### PR TITLE
Elbv2-Insecure Ciphers

### PR DESCRIPTION
**Pull request by alphadev4**

### CODE REVIEW

File: plugins/aws/elbv2/elbv2InsecureCiphers.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'ELBv2 Insecure Ciphers',
+    category: 'ELBv2',
+    domain: 'Content Delivery',
+    description: 'Ensure that Elbv2 listeners are configured to use the latest predefined security policies.',
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:8]: Is it correct?

File: plugins/aws/elbv2/elbv2InsecureCiphers.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    more_info: 'Insecure or deprecated security policies can expose the client and the load balancer to various vulnerabilities.',
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:9]: is it correct?

File: plugins/aws/elbv2/elbv2InsecureCiphers.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    link: 'https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html',
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:10]: Is it correct?

File: plugins/aws/elbv2/elbv2InsecureCiphers.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    recommended_action: 'Modify ELBv2 listeners with the latest predefined AWS security policies.',
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:11]: Is it correct?

File: plugins/aws/elbv2/elbv2InsecureCiphers.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            if (!describeLoadBalancers) return rcb();
+
+            if (describeLoadBalancers.err || !describeLoadBalancers.data) {
+                helpers.addResult(results, 3,
+                    `Unable to query for load balancers: ${helpers.addError(describeLoadBalancers)}`,region);
+                return rcb();
+            }
+
+            if (!describeLoadBalancers.data.length) {
+                helpers.addResult(results, 0, 'No load balancers found', region);
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:50]:
```suggestion
                helpers.addResult(results, 0, 'No Application or Network load balancers found', region);
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                        SslPolicy = true;
+                        if (deprecatedALBCipherPolicies.includes(listener.SslPolicy) || deprecatedNLBCipherPolicies.includes(listener.SslPolicy)) {
+                            depPolicies.push(listener.SslPolicy);
+                        }
+                    }
+                });
+                if (!SslPolicy){
+                    helpers.addResult(results, 0,'No SSL policies found', region, alb.LoadBalancerArn);
+                } else if (depPolicies && depPolicies.length){
+                    helpers.addResult(results, 2, `The listeners on "${alb.LoadBalancerName}" are using following policies which contains insecure ciphers ` + depPolicies.join(', '), region, alb.LoadBalancerArn);
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:83]:
```suggestion
                    helpers.addResult(results, 2, `Load balancer listeners have these policies with insecure ciphers: ` + depPolicies.join(', '), region, alb.LoadBalancerArn);
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                } else if (depPolicies && !depPolicies.length){
+                    helpers.addResult(results, 0, `All listeners on "${alb.LoadBalancerName}" are using policies which contains secure ciphers`, region, alb.LoadBalancerArn);
```
Comment (AkhtarAmir) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:85]:
```suggestion
                } else {
                    helpers.addResult(results, 0, 'Load balancer listeners policies contain secure ciphers', region, alb.LoadBalancerArn);
```

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/elbv2/elbv2InsecureCiphers.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+                var describeListeners = helpers.addSource(cache, source,
+                    ['elbv2', 'describeListeners', region, alb.DNSName]);
+
+                if (describeListeners.err || !describeListeners.data || !describeListeners.data.Listeners) {
+                    helpers.addResult(results, 3, `Unable to query for Listeners: ${helpers.addError(describeListeners)}`, region);
+                    continue;
+                }
+
+                if (!describeListeners.data.Listeners.length) {
+                    helpers.addResult(results, 0, 'No Listeners found', region, alb.LoadBalancerArn);
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:68]:
```suggestion
                    helpers.addResult(results, 0, 'No Listeners found for load balancer', region, alb.LoadBalancerArn);
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.spec.js [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+                expect(results[0].status).to.equal(3);
+                done();
+            });
+        });
+
+        it('should UNKNOWN if unable to describe Listeners', function (done) {
+            const cache = createCache([describeLoadBalancers[0]]);
+            elbv2InsecureCiphers.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('Unable to query for Listeners');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.spec.js:151]:
```suggestion
                expect(results[0].message).to.include('Unable to query for ELBv2 Listeners');
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.spec.js [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+                expect(results[0].status).to.equal(3);
+                done();
+            });
+        });
+
+        it('should PASS if no Listeners found', function (done) {
+            const cache = createCache([describeLoadBalancers[0]],describeListeners[2]);
+            elbv2InsecureCiphers.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('No Listeners found');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.spec.js:161]:
```suggestion
                expect(results[0].message).to.include('No Listeners found for load balancer');
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.spec.js [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+describe('elbv2InsecureCiphers', function () {
+    describe('run', function () {
+
+        it('should PASS if No Application Load Balancers found', function (done) {
+            const cache = createCache([]);
+            elbv2InsecureCiphers.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('No load balancers found');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.spec.js:131]:
```suggestion
                expect(results[0].message).to.include('No Application or Network load balancers found');
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.spec.js [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+                expect(results[0].status).to.equal(0);
+                done();
+            });
+        });
+
+        it('should PASS if no SSL policies found', function (done) {
+            const cache = createCache([describeLoadBalancers[0]],describeListeners[1]);
+            elbv2InsecureCiphers.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('No SSL policies found');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.spec.js:171]:
```suggestion
                expect(results[0].message).to.include('No SSL policies found for load balancer');
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.spec.js [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+                expect(results[0].status).to.equal(0);
+                done();
+            });
+        });
+
+        it('should PASS if listeners on Elbv2 are using policies which support insecure ciphers', function (done) {
+            const cache = createCache([describeLoadBalancers[0]],describeListeners[0]);
+            elbv2InsecureCiphers.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('are using following policies which contains insecure ciphers');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.spec.js:181]:
```suggestion
                expect(results[0].message).to.include('Load balancer listeners policies contain secure ciphers');
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.spec.js [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+                expect(results[0].status).to.equal(2);
+                done();
+            });
+        });
+
+        it('should PASS if all listeners on Elbv2 are using current policies', function (done) {
+            const cache = createCache([describeLoadBalancers[0]],describeListeners[3]);
+            elbv2InsecureCiphers.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('are using policies which contains secure ciphers');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.spec.js:191]:
```suggestion
                expect(results[0].message).to.include('Load balancer listeners have these policies with insecure ciphers');
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+                        SslPolicy = true;
+                        if (deprecatedALBCipherPolicies.includes(listener.SslPolicy) || deprecatedNLBCipherPolicies.includes(listener.SslPolicy)) {
+                            depPolicies.push(listener.SslPolicy);
+                        }
+                    }
+                });
+                if (!SslPolicy){
+                    helpers.addResult(results, 0,'No SSL policies found for load balancer', region, alb.LoadBalancerArn);
+                } else if (depPolicies && depPolicies.length){
+                    helpers.addResult(results, 2, `Load balancer listeners have these policies with insecure ciphers: ` + depPolicies.join(', '), region, alb.LoadBalancerArn);
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:83]:
```suggestion
                    helpers.addResult(results, 2, `Load balancer listeners have these policies with insecure ciphers: ` + depPolicies.join(', '), region, alb.LoadBalancerArn);
```
```suggestion
                    helpers.addResult(results, 2, 'Load balancer listeners have these policies with insecure ciphers: ' + depPolicies.join(', '), region, alb.LoadBalancerArn);
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.spec.js [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+                expect(results[0].status).to.equal(0);
+                done();
+            });
+        });
+
+        it('should PASS if listeners on Elbv2 are using policies which support insecure ciphers', function (done) {
+            const cache = createCache([describeLoadBalancers[0]],describeListeners[0]);
+            elbv2InsecureCiphers.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('Load balancer listeners policies contain secure ciphers');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.spec.js:181]:
```suggestion
                expect(results[0].message).to.include('Load balancer listeners have these policies with insecure ciphers');
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.spec.js [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+                expect(results[0].status).to.equal(2);
+                done();
+            });
+        });
+
+        it('should PASS if all listeners on Elbv2 are using current policies', function (done) {
+            const cache = createCache([describeLoadBalancers[0]],describeListeners[3]);
+            elbv2InsecureCiphers.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('Load balancer listeners have these policies with insecure ciphers');
```
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.spec.js:191]:
```suggestion
                expect(results[0].message).to.include('Load balancer listeners policies contain secure ciphers');
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.js (hunk 1 above)
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:8]:
```suggestion
    description: 'Ensure that Elbv2 listeners are configured to use the predefined security policies containing secure ciphers.',
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.js (hunk 3 above)
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:10]:
```suggestion
    link: 'https://docs.aws.amazon.com/elasticloadbalancing/latest/network/create-tls-listener.htmll',
```

File: plugins/aws/elbv2/elbv2InsecureCiphers.js (hunk 4 above)
Comment (alphadev4) [inline plugins/aws/elbv2/elbv2InsecureCiphers.js:11]:
```suggestion
    recommended_action: 'Modify ELBv2 listeners with the predefined AWS security policies containing secure ciphers.',
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**